### PR TITLE
Layermanager now tracks all EditTargets

### PR DESCRIPTION
**Pull request by dbaz**
## Description
@nxkb @elrond79 @sirpalee 

This is a preview of some code I modified during a task I'm part way through for some AL internal pipeline work where I need to be able to get all the layers that have been modified so I can serialise them out.

Previously, the layerManager wasn't keeping track of the first EditTarget or the last EditTarget that was set, only the previously set EditTarget. Now, if you set an EditTarget the LayerManager immediately starts to track the layer.

The LayerManager also now has the option to only return "Dirty" layers, which are layers that have changed from it's persistent(serialised?) representation. The API defaults to only return the "Dirty" layers.

Let me know if this logic breaks anything on your ends.

### Added

- You can now ask for all layers or just the dirty layers from the LayerManager

### Changed

- All layers that are set as the EditTarget are tracked by the LayerManager

## Checklist (Please do not remove this line)
- [ ] Make sure the Title and Description of the PR make sense and  provide sufficient context for your work
- [ ] Do any added files have the correct AL Apache Licence Header?
- [ ] Are there Doxygen comments in the headers?
- [ ] Are any new features, behaviour changes documented in the .md format [documentation](https://github.com/AnimalLogic/AL_USDMaya/docs)
- [ ] Have you added, updated tests to cover new features and behaviour changes
- [ ] Do all the commits in this PR have a meaningful message?
- [ ] Can ChangeLog info be extracted from the PR description or commit messages

### CODE REVIEW

**Comment by dbaz** [comment]
tldr;
Yes, I just need to know about all the layers that have been dirtied. No, I don't need to know about all the Layers that have been EditTarget.

If I call ```trackEditTargetLayer()``` just before wanting to grab the dirty EditTargets (or add that flag you mentioned), in it's current implementation, will track the current EditTarget's layer even if it hasn't been dirtied. Does it follow the design to put the logic to track only dirty layers here? 


Bit longer brain dump:
I see the paradigm is shifting in this way:
A) LayerManager will only track layers that were dirtied. Is up to be up-to-date when serialising
B) LayerManager will track all layers that have been set as the EditTarget. Is updated when EditTarget is set.


The two solutions are a bit different in the way of "how and when" LayerManager is up to date with the dirty layers. Previously, it was guarentee'd up to date when saving the scene, but I wanted it to be up to date immediatly after doing edits on a newly set EditTarget. I have a usecase lower down.


```Consider the scenario where we open a maya file which has a serialized layer, and then immediately save it again. It's not hard to imagine that a layer might be considered not dirty```
I'm not sure how my change could effect the previous logic with the deserialisation, but from my tests the ``ImportFromString`` marks it as dirty, so it will be saved again, in both implementations.


```Taking a step back - it generally seems a safer approach to say:

A) only save in our local storage things that we really need to... but then once it's there, it STAYS there until we explicitly remove it```

Yep this is the optimal solution, to track only layers that are 'dirty' and that have been the EditTarget.
I agree that my solution isn't optimal because as you stated since it tracks all layers that have been the EditTarget, even if it isn't 'dirtied', though I don't think it's a performance killer or anything like that, but it can be a tad wasteful. To do this optimally is a bit more work using USD's notices. But all this I think is outside the original design of the LayerManager who's main goal was the serialise out all the dirty layers.

```
Did you actually run into a scenario where this was the case, or is this just based on your reading of the code?
```
Yep, I ran through the usecase below.

```However, it is not just called by the onEditTargetChanged() callback; it is also called when the stage is first created, in ProxyShape::loadStage(), and immediately before it is serialized, in ProxyShape::serialize()```

This is where I ran into problems, I don't save the scene in my use case, but yes, it correctly serialises all the layers that have been edited when saved, but I wanted to be able to ask the LayerManager for all the edited layers at any time. 

So the steps I do are:
1 Load a ProxyShape
2 Set the EditTarget to my SubLayer
3 Modify the EditTarget contents
4 Run a script that exports the Maya scene into our renderer's scene description

During step 3, no LayerManager node is created, so when the script runs to export our renderer's scene description it can't find any modified layers.

**Comment by elrond79** [comment]
**tldr;**
Though both approaches would work, I would feel safer keeping the current paradigm for LayerManager, and just adding a call to be able to query all the modified layers (that will update with all current edit targets).  I'll submit a PR that takes this approach, so you can take a look and see if it meets your needs.

**Longer version:**

 > If I call trackEditTargetLayer() just before wanting to grab the dirty EditTargets (or add that flag you mentioned), in it's current implementation, **will track the current EditTarget's layer even if it hasn't been dirtied.** 

Hmm... I don't think this is true. If you look at the code for `trackEditTargetLayer`, the addLayer call is protected behind a check if the layer is dirty:

```
  if(prevTargetLayer->IsDirty())
  {
    if(!layerManager)
    {
        ...
    }
    layerManager->addLayer(prevTargetLayer);
  }
```

(Note that in the case where we're calling it "manually", ie not as a result of the editTargetChanged callback, prevTargetLayer will be the same as the current edit target.)

> Does it follow the design to put the logic to track only dirty layers here?

As I mentioned - it currently only tracks dirty layers. The caveat is that it's not currently guaranteed to track ALL dirty layers... ie, without the extra call to trackEditTargetLayer, it may not track the current edit target (which may be dirty).

>>  Taking a step back - it generally seems a safer approach to say:
>>  A) only save in our local storage things that we really need to... but then once it's there, it STAYS there until we explicitly remove it
> Yep this is the optimal solution, to track only layers that are 'dirty' and that have been the EditTarget.
> I agree that my solution isn't optimal because as you stated since it tracks all layers that have been the EditTarget, even if it isn't 'dirtied', though I don't think it's a performance killer or anything like that, but it can be a tad wasteful. 

I'm not too worried about performance / optimization - I'm concerned about data safety / data loss. Perhaps another way to frame the high-level view would be this:

A) Data to be saved is put in a safe.  Before anything is put in the safe, it is checked to make sure it is something we want to save.  Everything in the safe is always saved, and once something gets in the safe, it is never removed, unless explicitly requested by the user.

B) Data to be saved is put in a sieve.  When it comes time to save, everything in the sieve is filtered, and only the filtered results that come out of the bottom of the sieve are saved.  This filtering is done every time we save, so even results that previously passed through the sieve, and were saved, are re-filtered to see if they still pass through the filter, before they are re-saved.

The key difference is here, in LayerManager::populateSerializationAttributes:

https://github.com/AnimalLogic/AL_USDMaya/blob/fe2e914b2fc8625ccb00fdb99e7891d80a8d22e4/lib/AL_USDMaya/AL/usdmaya/nodes/LayerManager.cpp#L439

```
MStatus LayerManager::populateSerialisationAttributes()
{
  TF_DEBUG(ALUSDMAYA_LAYERS).Msg("LayerManager::populateSerialisationAttributes\n");
  const char* errorString = "LayerManager::populateSerialisationAttributes";
  ...
  {
    boost::shared_lock_guard<boost::shared_mutex> lock(m_layersMutex);
    ...
    for (const auto& layerAndIds : m_layerDatabase)
    {
      auto& layer = layerAndIds.first;
      if(layer->IsDirty())
      {
```

This is the line that checks if a layer is dirty before it is allowed to be saved, and the key difference is that it will be applied to layers which have already been saved.

Now, as you've noted, due to the fact that ImportFromString marks layers as dirty, and we're currently always loading layers using that call, saved layers should always stay saved.  And I will admit it's unlikely, but what if either of those preconditions change?  I guess it just makes me feel safer to design the logic of the program to ensure data safety, rather than relying on implementation details.

> This is where I ran into problems, I don't save the scene in my use case, but yes, it correctly serialises all the layers that have been edited when saved, but I wanted to be able to ask the LayerManager for all the edited layers at any time.
> 
> So the steps I do are:
> 1 Load a ProxyShape
> 2 Set the EditTarget to my SubLayer
> 3 Modify the EditTarget contents
> 4 Run a script that exports the Maya scene into our renderer's scene description
>
> During step 3, no LayerManager node is created, so when the script runs to export our renderer's scene description it can't find any modified layers.

Ok, this makes sense.  However, I'd feel more comfortable if we could solve this by just modifying step 4, so that it makes an extra call to ensure the LayerManager is "up to date", rather than changing the LayerManager paradigm from A) to B).  I'll make a PR that adds methods for updating the LayerManager with the "latest" changes, and a method that returns all modified layers.

**Comment by elrond79** [comment]
Hey... so I was thinking about this some more, and I'm coming around to your point of view.  Being able to always know you're querying the complete current set of modified layers, without having to worry about calling some sort of update function first, is a pretty big improvement... and if the only argument against is a purely theoretical one, that's probably not worth sacrificing that feature.  Plus, there are other ways to ensure previously-saved data is always preserved (ie, add a boolean flag).

If you really only do care about dirty layers, though, I may change the interface to get rid of the dirtyOnly flags...

**Comment by elrond79** [comment]
Made some commits, but due to some technical difficulties at our studio right now, I can't push commits at the moment.

Basically, I altered the interface to LayerDatabase so that you can add non-dirty layers to it, but they're essentially hidden - queries won't find them, and it won't return them when iterating over it's contents.  However, as soon as they become dirty, they will magically show up.

Also, made it so that when the edit target is changed, if the previous edit target still isn't dirty, it is removed from the LayerManger, so we avoid cluttering it up with useless non-dirty layers; at any given point, if you have N proxy shapes in the scene, there should be at most N "hidden" non-dirty layers in the LayerManager.

Server issues should hopefully be resolved by our Monday, so I'll push them then.

**Comment by dbaz** [comment]
Hey @elrond79 could you push up your branch for PR when you get the chance, I want to get it merged so I can build some of my work on top of it. Thanks!

**Comment by elrond79** [comment]
Done, over here:
https://github.com/AnimalLogic/AL_USDMaya/pull/84

Sorry, we're still having problems with the conduit to the outside world (we operate in a closed shop, without access to the internet, so getting stuff out is a bit complicated).  On the plus side, I learned about "git bundle"...

**Comment by dbaz** [comment]
Yeah it's Odd. Anyways, I've pulled your main commits locally to AL and one of us will be pushing an AL_USDMaya release mid next week. Closing this PR. Thanks!
